packages/liveblocks-client/src/ImmRef.ts

```diff
@@ -0,0 +1,67 @@
+import type { EventSource, Observable } from "./EventSource";
+import { makeEventSource } from "./EventSource";
+
+/**
+ * Patches a target object by "merging in" the provided fields. Patch
+ * fields that are explicitly-undefined will delete keys from the target
+ * object. Will return a new object.
+ *
+ * Important guarantee:
+ * If the patch effectively did not mutate the target object because the
+ * patch fields have the same value as the original, then the original
+ * object reference will be returned.
+ */
+export function merge<T>(target: T, patch: Partial<T>): T {
+  let updated = false;
+  const newValue = { ...target };
+
+  Object.keys(patch).forEach((k) => {
+    const key = k as keyof T;
+    const val = patch[key];
+    if (newValue[key] !== val) {
+      if (val === undefined) {
+        delete newValue[key];
+      } else {
+        newValue[key] = val as T[keyof T];
+      }
+      updated = true;
+    }
+  });
+
+  return updated ? newValue : target;
+}
+
+/**
+ * Base class that implements an immutable cache.
+ *
+ * TODO: Document usage.
+ */
+export abstract class ImmRef<T> {
```

Comment:
```suggestion
export abstract class ImmutableRef<T> {
```

Reply:
Can update! I'm still looking for a slightly better abstraction here. I still feel like this abstract class is a little too clunky/bulky to use and am still looking for a slightly lighter API to enable this, but I can do that in a follow-up PR to clean that up. For now, it works.

Until I found the better abstraction that's still itching, I'll change this name — thanks 👍 

Done in 4607112d8e367a48323835d027064e3c6cadae08.